packages/jspsych/src/modules/data/index.ts

```diff
@@ -59,8 +59,10 @@ export class JsPsychData {
   write(trial: Trial) {
     const result = trial.getResult();
     Object.assign(result, this.dataProperties);
-    this.results.push(result);
-    this.resultToTrialMap.set(result, trial);
+    if (trial.getParameterValue("record_data") ?? true) {
```

Comment:
This breaks at https://github.com/jspsych/jsPsych/blob/6f9d01b2ae0a35f38fadf78d8311cf501121681e/packages/jspsych/src/timeline/Timeline.ts#L96 where `record_data` is not respected. What do you think about integrating it at the trial level in `getResult`/`getResults`? [Timeline-level getResults()](https://github.com/jspsych/jsPsych/blob/6f9d01b2ae0a35f38fadf78d8311cf501121681e/packages/jspsych/src/timeline/Timeline.ts#L240) already filters out undefined results too :ok_hand:

Reply:
I added a test case for this and got it to pass by having getResult(s) return undefined when record_data is false, but it looks like there are other parts of the new core logic that expect getResult() to return an object. There are some failing tests now. I'm not sure what the best solution is since I'm not totally comfortable with this part of the new core. Any suggestions?

Reply:
Oh, the problem with this is
https://github.com/jspsych/jsPsych/blob/33c97c242329dd68370388648cafa7a61cda2152/packages/jspsych/src/JsPsych.ts#L348
in `onTrialResultAvailable()`. While looking into this, I noticed jsPsych's global `on_trial_finish` callback would also be invoked with `undefined` with this approach. Which is likely not what we want, is it? In this sense, maybe the loop function should also get all results, including those which are not recorded (just like your initial solution)?

Reply:
I'm thinking about #2736 in this context. Maybe it makes sense to keep `on_data_update()`, which only gets triggered when there is a change to the data object, but allow `on_trial_finish()` and the trial's `on_finish()` methods to be called if there is no data. I'm not sure if `undefined` or an empty DataCollection is the best thing for those methods to receive. I don't think they should get the data from the trial if `record_data` is false, since that would mean different kinds of filtering methods in different contexts.

Reply:
> I don't think they should get the data from the trial if record_data is false, since that would mean different kinds of filtering methods in different contexts.

I agree. Then `on_trial_finish()`  and `on_data_update()` both make sense again :+1: